File: sdk-rust/src/bin/migrate.rs
// Copyright (c) 2026 12vectors.com
// SPDX-License-Identifier: MIT
// See the LICENSE file in the repository root for the full license text.

//! Standalone migration runner.
//!
//! Usage:
//!     cargo run --bin gatedhouse-migrate -- <conninfo>

use std::process::ExitCode;
use std::sync::Arc;

use gatedhouse::{ConninfoDatabase, Database, GatedhouseConfig, GatedhouseFactory};

fn main() -> ExitCode {
    let args: Vec<String> = std::env::args().skip(1).collect();
    if args.len() != 1 {
        eprintln!(
            "Usage: gatedhouse-migrate <conninfo>\n\n\
             Example:\n\
             \x20   gatedhouse-migrate 'host=localhost user=postgres password=secret dbname=mydb'"
        );
        return ExitCode::from(2);
    }

    let database: Arc<dyn Database> = Arc::new(ConninfoDatabase::new(&args[0]));
    let config = GatedhouseConfig::builder(database).build();

    match GatedhouseFactory::migrate(&config) {
        Ok(()) => {
            println!("Gatedhouse migration completed successfully.");
            ExitCode::SUCCESS
        }
        Err(e) => {
            eprintln!("Gatedhouse migration failed: {e}");
            ExitCode::FAILURE
        }
    }
}

File: sdk-java/src/main/java/com/twelvevectors/gatedhouse/cli/Migrate.java
// Copyright (c) 2026 12vectors.com
// SPDX-License-Identifier: MIT
// See the LICENSE file in the repository root for the full license text.

package com.twelvevectors.gatedhouse.cli;

import com.twelvevectors.gatedhouse.Database;
import com.twelvevectors.gatedhouse.GatedhouseConfig;
import com.twelvevectors.gatedhouse.GatedhouseFactory;

public final class Migrate {

    private Migrate() {
    }

    public static void main(String[] args) {
        if (args.length < 2 || args.length > 3) {
            System.err.println(
                "Usage: java -cp gatedhouse-<version>.jar:postgresql-<version>.jar \\\n"
                + "         com.twelvevectors.gatedhouse.cli.Migrate \\\n"
                + "         <jdbc-url> <user> [password]\n"
                + "\n"
                + "Example:\n"
                + "  java -cp gatedhouse.jar:postgresql.jar com.twelvevectors.gatedhouse.cli.Migrate \\\n"
                + "       jdbc:postgresql://localhost:5432/mydb gatedhouse_user secret"
            );
            System.exit(2);
        }

        String url = args[0];
        String user = args[1];
        String password = args.length == 3 ? args[2] : "";

        Database database = Database.fromUrl(url, user, password);
        GatedhouseConfig config = GatedhouseConfig.builder()
            .database(database)
            .build();

        try {
            GatedhouseFactory.migrate(config);
            System.out.println("Gatedhouse migration completed successfully.");
        } catch (RuntimeException e) {
            System.err.println("Gatedhouse migration failed: " + e.getMessage());
            Throwable cause = e.getCause();
            if (cause != null) {
                System.err.println("Caused by: " + cause);
            }
            System.exit(1);
        }
    }
}
